### PR TITLE
Support evaluation on VisDial test split.

### PR DESCRIPTION
**Pull request by kdexd**
This PR makes changes in **evaluate.lua** and underlying modifications to dataloader and model files to perform evaluation on test split of VisDial. The schema of test split is described in description of #16 

100 candidate answer options of the dialog round are ranked according to the likelihood in absence of ground truth answer (in test split). The retrieved ranks table in **evaluate.lua** can be saved as a json file (list of json objects, where each object has three keys:

```text
{
    'image_id': 823457, // integer image id from the dataset
    'ranks' : [....], // list of 100 integers, n-th number is rank of n-th candidate answer option
    'round_id': 4 // any integer between 1 to 10
}
```

* For the val split, to get such JSON, `-useGt` flag must be set to false -- setting it to true uses the ground truth answers from the dataset and hence `ranks` comes out to be a single integer, meaning the rank of ground truth. `-UseGt` is set to false forcefully for test split.

* There shall be one such JSON object per image_id for test split, since only ranks of last dialog round are retrieved, whereas there shall be ten such JSON objects for val split, since ranks of all rounds are retrieved.

### CODE REVIEW

File: model.lua [hunk 1]
```diff
@@ -150,33 +150,57 @@ function Model:retrieve(dataloader, dtype)
...
-        -- grab a batch
-        local batch, nextStartId =
-                        dataloader:getTestBatch(startId, self.params, dtype);
-
-        -- Call retrieve function for specific model, and store ranks
-        ranks[{{startId, nextStartId - 1}, {}}] = self:retrieveBatch(batch);
-        startId = nextStartId;
-    end
+    local retrievalOutput;
+    if dtype == 'val' then
```
Comment (satwikkottur) [inline model.lua]: Should this be for both 'val' and 'train', just in case someone wants to compute retrieval metrics on training set?

File: model.lua [hunk 2]
```diff
@@ -150,33 +150,57 @@ function Model:retrieve(dataloader, dtype)
...
+
+            -- grab a batch
+            local batch, nextStartId =
+                            dataloader:getTestBatch(startId, self.params, dtype);
+
+            -- Call retrieve function for specific model, and store ranks
+            local batchAnswerOptions = self:retrieveBatch(batch, dtype);
+            -- remove dummy answer options for first nine rounds
+            batchAnswerOptions = batchAnswerOptions:view(nextStartId - startId, -1, self.params.numOptions)
+            answerOptions[{{startId, nextStartId - 1}, {}}] = batchAnswerOptions[{{}, {10}}];
```
Comment (satwikkottur) [inline model.lua]: Unless I am wrong, shouldn't you be picking the round based on the round for the current test instance? I remember we padding zeros to later part of the dialog. For instance, if the test instance is at round 5, we pad rounds 6..10 as 0, and need to pick the rank at round 5, correct?

File: model.lua (hunk 1 above)
Comment (kdexd) [reply model.lua]: Okay that makes sense, I'll do `dtype ~= 'test'`

File: model.lua (hunk 2 above)
Comment (kdexd) [reply model.lua]: @satwikkottur I guess you reviewed on an outdated diff. Please review on the latest commit - there are two methods `retrieve` and `predict`. 
While this line stays, the ranks list of last round is picked out while forming the retrieval table (which is ready to be saved as JSON).
This line in particular saves the rank list at `round_id` index of ranks table: https://github.com/batra-mlp-lab/visdial/pull/18/commits/83cfa1f86d1aeb5a3bd8f9446d7014600535637e#diff-cd27003c24a2c135a29b100aa893b075R236

File: model.lua (hunk 1 above)
Comment (kdexd) [reply model.lua]: Oh, this is on an outdated diff too -- all these changes reflect when I wrote the code according to my first approach. I changed it to separate methods and this line got lost. Although I will make a minor change in upcoming commit.

File: model.lua [hunk 3]
```diff
@@ -222,20 +222,20 @@ function Model:predict(dataloader, dtype)
     local ranks = torch.totable(ranks:double());
     for i = 1, #dataloader['unique_img_'..dtype] do
         -- rank list for all rounds in val split and last round in test split
-        if dtype == 'val' then
+        if dtype == 'test' then
```
Comment (kdexd) [inline model.lua]: @satwikkottur This is similar to what you reviewed before, I hope this makes sense if someone needs to retrieve on train split itself.

File: utils.lua [hunk 4]
```diff
@@ -104,9 +104,19 @@ end
...
+    -- simply sort according to scores if ground truth not available
+    local ranks;
+    if gtPos then
+        gtPos = gtPos:view(-1, 1);
+        local gtScore = scores:gather(2, gtPos);
+        ranks = scores:gt(gtScore:expandAs(scores));
+        ranks = ranks:sum(2) + 1;
+    else
+        sorted, ranks = scores:sort(2)
+        for i = 1, scores:size(1) do
```
Comment (satwikkottur) [inline utils.lua]: I am not sure if this is doing the right thing. A unit test below with the output.

``` 
  1 require 'torch'
  2
  3 scores = torch.FloatTensor(10, 3):uniform();
  4
  5 sorted, ranks = scores:sort(2)
  6     for i = 1, scores:size(1) do
  7         ranks[i] = ranks[i]:index(1, ranks[i]);
  8     end
  9 print(scores)
 10 print(ranks)```

Output:
 0.1847  0.9904  0.1582
 0.6934  0.3652  0.8853
 0.5281  0.3054  0.7413
 0.5424  0.2891  0.4829
 0.2586  0.3230  0.8476
 0.6061  0.9546  0.3052
 0.5951  0.8214  0.7093
 0.1442  0.8960  0.9145
 0.6027  0.2672  0.0815
 0.3776  0.3473  0.5396
[torch.FloatTensor of size 10x3]

 2  3  1
 1  2  3
 1  2  3
 3  1  2
 1  2  3
 2  3  1
 1  2  3
 1  2  3
 1  2  3
 1  2  3
[torch.LongTensor of size 10x3]

Comment (kdexd) [reply utils.lua]: Oh yes, the order should be reversed.

Comment (kdexd) [reply utils.lua]: Thanks for pointing it out @satwikkottur, here is the test:

```
  1 require 'torch'
  2
  3 scores = torch.FloatTensor(10, 3):uniform();
  4
  5 sorted, ranks = scores:sort(2, true)
  6 print(scores)
  7 print(ranks)
```
```
 0.2974  0.6810  0.9164
 0.9497  0.2618  0.0178
 0.1139  0.3085  0.0496
 0.3462  0.6437  0.9521
 0.8896  0.7418  0.7638
 0.8688  0.5988  0.3765
 0.5683  0.8200  0.8018
 0.2770  0.3476  0.8543
 0.7989  0.4341  0.3507
 0.1155  0.4421  0.8378
[torch.FloatTensor of size 10x3]

 3  2  1
 1  2  3
 2  1  3
 3  2  1
 1  3  2
 1  2  3
 2  3  1
 3  2  1
 1  2  3
 3  2  1
[torch.LongTensor of size 10x3]
```
